openedx/core/djangoapps/schedules/management/commands/tests/test_send_recurring_nudge.py

```diff
@@ -41,63 +40,72 @@ def setUp(self):
         mock_resolver.assert_called_with(self.site_config.site, test_time)
 
-        for day in (3, 10):
+        for day in (-3, -10):
             mock_resolver().send.assert_any_call(day, None)
 
     @patch.object(tasks, 'ace')
-    @patch.object(nudge, 'recurring_nudge_schedule_hour')
-    def test_resolver_send(self, mock_schedule_hour, mock_ace):
+    @patch.object(resolvers.ScheduleStartResolver, 'async_send_task')
+    def test_resolver_send(self, mock_schedule_bin, mock_ace):
         current_time = datetime.datetime(2017, 8, 1, tzinfo=pytz.UTC)
-        nudge.ScheduleStartResolver(self.site_config.site, current_time).send(3)
-        test_time = current_time - datetime.timedelta(days=3)
-        self.assertFalse(mock_schedule_hour.called)
-        mock_schedule_hour.apply_async.assert_any_call(
-            (self.site_config.site.id, 3, serialize(test_time), [], True, None),
+        nudge.ScheduleStartResolver(self.site_config.site, current_time).send(-3)
+        test_time = current_time + datetime.timedelta(days=-3)
+        self.assertFalse(mock_schedule_bin.called)
+        mock_schedule_bin.apply_async.assert_any_call(
+            (self.site_config.site.id, serialize(test_time), -3, 0, [], True, None),
             retry=False,
         )
-        mock_schedule_hour.apply_async.assert_any_call(
-            (self.site_config.site.id, 3, serialize(test_time + datetime.timedelta(hours=23)), [], True, None),
+        mock_schedule_bin.apply_async.assert_any_call(
+            (self.site_config.site.id, serialize(test_time), -3, tasks.RECURRING_NUDGE_NUM_BINS - 1, [], True, None),
             retry=False,
         )
         self.assertFalse(mock_ace.send.called)
 
     @ddt.data(1, 10, 100)
     @patch.object(tasks, 'ace')
     @patch.object(tasks, '_recurring_nudge_schedule_send')
-    def test_schedule_hour(self, schedule_count, mock_schedule_send, mock_ace):
+    def test_schedule_bin(self, schedule_count, mock_schedule_send, mock_ace):
         schedules = [
-            ScheduleFactory.create(start=datetime.datetime(2017, 8, 1, 18, 34, 30, tzinfo=pytz.UTC))
-            for _ in range(schedule_count)
+            ScheduleFactory.create(
+                start=datetime.datetime(2017, 8, 3, 18, 44, 30, tzinfo=pytz.UTC),
+                enrollment__user=UserFactory.create(),
+                enrollment__course__id=CourseLocator('edX', 'toy', 'Bin')
+            ) for _ in range(schedule_count)
         ]
 
-        test_time_str = serialize(datetime.datetime(2017, 8, 1, 18, tzinfo=pytz.UTC))
-        with self.assertNumQueries(2):
-            tasks.recurring_nudge_schedule_hour(
-                self.site_config.site.id, 3, test_time_str, [schedules[0].enrollment.course.org],
-            )
+        test_time = datetime.datetime(2017, 8, 3, 18, tzinfo=pytz.UTC)
+        test_time_str = serialize(test_time)
+        with self.assertNumQueries(25):
```

Comment:
I think it might be better to have the `assertNumQueries` inside the bins loop, so that it's clearer how many queries each bin uses.

Reply:
I add to do some weirdness around the extra waffle_flag query that happens before the first schedules query, but it's inside the bin loop now.